Blog Post: Introducing JobSet (#45759)
kubernetes/website: content/en/blog/_posts/2024-04-02-introducing-jobset/2024-04-02-introducing-jobset.md

Reviewer: Please wrap the markdown at 100 or so (source) characters per line.

Once you've done that, I recommend squashing commits.

Reviewer 2: @danielvegamyhre has this been addressed?

Author: Wrapped at 100 characters

---
layout: blog
title: "Introducing JobSet"
date: 2024-04-02
slug: introducing-jobset
---

**Authors**: Daniel Vega-Myhre (Google), Abdullah Gharaibeh (Google), Kevin Hannon (Red Hat)

> In this article, we introduce [JobSet](https://jobset.sigs.k8s.io/), an open source API for representing distributed jobs. The goal of JobSet is to provide a stable API for running/building APIs with AI/ML and HPC use cases in mind.

Reviewer 3: I struggle if we should define HPC..

What do you mean by "running/building APIs"?

Reviewer 2: I think HPC is a known term, but I also don't understand the running/building APIs part.

> 

Reviewer: Please mention that it is alpha. The goal might be stability, but “provide a stable API” might mislead people into thinking it already is.

Reviewer 2: Lets just say "provide an API", we don't need to mention stable or alpha as it will eventually graduate.

Author: Updated phrasing, please let me know what you think.

Reviewer: This is fine.

## Why JobSet?

> The Kubernetes community’s recent enhancements to the batch ecosystem on Kubernetes attracted ML engineers who have found it to be a natural fit for the requirements of running distributed training workloads. Large ML models (particularly LLMs) which cannot fit into the memory of the GPU or TPU chips on a single host are often distributed across tens of thousands of accelerator chips, which in turn may span thousands of hosts. 

Reviewer 3: Please add some new lines. (I can only comment on this entire paragraph).

Reviewer 3: With this, I try to follow a newline per sentence.

Author: Hmm this paragraph is only 2 sentences. Do we really need/want each sentence to be it's own paragraph? In the Kueue blog post they use paragraphs: https://kubernetes.io/blog/2022/10/04/introducing-kueue/

Reviewer 3: I think it will be rendered the same. But if you have no new lines, people can only suggest the entire paragraph.

Reviewer: Localization teams ask for Markdown where the source is wrapped at around 100 characters.

Please do that.

> As such, the model training code is often containerized and executed simultaneously on all these hosts, performing distributed computations which often shard both the model parameters and/or the training dataset across the target accelerator chips, using NCCL or XLA collective primitives like all-gather and all-reduce to perform distributed computations and synchronize gradients between hosts. 

Reviewer 3: Please define definitions. NCCL or XLA is well known enough for the general audience.

Author: How about making NCCL and XLA links to the docs? 

Otherwise the sentence may become a bit hard to follow:

`... using NCCL (library implementing multi-GPU and multi-node communication primitives) or XLA (compiler targeting XLA devices, such as TPUs, including collective primitives for multi-TPU and multi-node communication) collective primitives like all-gather and all-reduce to perform...`

Author: True, maybe it can just be:

`... using collective primitives like all-gather and all-reduce to perform...`

Reviewer 3: SGTM.

These workload characteristics make Kubernetes a great fit for this type of workload, as efficiently scheduling and managing the lifecycle of containerized applications across a cluster of compute resources is an area where it shines. It is also very extensible, allowing developers to define their own Kubernetes APIs, objects, and controllers which manage the behavior and life cycle of these objects, allowing engineers to develop custom distributed training orchestration solutions to fit their needs.

However, as distributed ML training techniques continue to evolve, existing Kubernetes primitives do not adequately model them alone anymore. Furthermore, the landscape of Kubernetes distributed training orchestration APIs has become fragmented, and each of the existing solutions in this fragmented landscape has certain limitations that make it non-optimal for distributed ML training. 

> On the one hand, the KubeFlow training operator defines custom APIs for different ML frameworks (e.g. PyTorchJob, TFJob, MPIJob, etc.); however, each of these job types are in fact a bespoke solution fit specifically to the target framework, each with different semantics and behavior. 

Reviewer 3: ```suggestion
On the one hand, the KubeFlow training operator defines custom APIs for different ML frameworks (e.g. PyTorchJob, TFJob, MPIJob, etc.); however, each of these job types are in fact a solution fit specifically to the target framework, each with different semantics and behavior. 
```

Bespoke reads like a put down of the project. it fits a need but we shouldn't cast judgement.

On the other hand, the Job API fixed many gaps for running batch workloads, including Indexed completion mode, higher scalability, Pod failure policies and Pod backoff policy to mention a few of the most recent enhancements. However, running ML training and HPC workloads using the upstream Job API requires extra orchestration to fill the following gaps:

> 1) **Multi-template Pods**: Most HPC or ML training jobs include more than one type of Pods. The different Pods are part of the same workload, but they need to run a different container, request different resources or have different failure policies. A common example is the driver-worker pattern.
> 
> 2) **Job Groups**: Large scale training workloads span multiple network topologies, running across multiple racks for example. Such workloads are network latency sensitive, and aim to localize communication and minimize traffic crossing the higher-latency network links. To facilitate this, the workload needs to be split into groups of Pods each assigned to a network topology.
> 
> 3) **Network setup**: Create and manage the resources (e.g. headless services) necessary to establish communication between the Pods of a job. 
> 
> 4) **Startup sequence**: Some jobs require a specific start sequence of pods; sometimes the driver is expected to start first (like Ray or Spark), in other cases the workers are expected to be ready before starting the driver (like MPI). 

Reviewer: Try using a description list. Here's what I'd write (this Markdown renders to a `<dl>` element):

```suggestion
Multi-template Pods
: Most HPC or ML training jobs include more than one type of Pods. The different Pods are part of the same workload, but they need to run a different container, request different resources or have different failure policies. A common example is the driver-worker pattern.

Job groups
: Large scale training workloads span multiple network topologies, running across multiple racks for example. Such workloads are network latency sensitive, and aim to localize communication and minimize traffic crossing the higher-latency network links. To facilitate this, the workload needs to be split into groups of Pods each assigned to a network topology.

Inter-Pod communication
: Create and manage the resources (e.g. [headless Services](/docs/concepts/services-networking/service/#headless-services)) necessary to establish communication between the Pods of a job. 

Startup sequencing
: Some jobs require a specific start sequence of pods; sometimes the driver is expected to start first (like Ray or Spark), in other cases the workers are expected to be ready before starting the driver (like MPI).
```

As an exception, these are easiest to work with if you put the description all on one line.

Author: Updated this section (and others) to use description list format

5) **Integration with Kueue**: users can submit JobSets via [Kueue](https://kueue.sigs.k8s.io/) to oversubscribe their clusters, queue workloads to run as capacity becomes available, prevent partial scheduling and deadlocks, enable multi-tenancy, and more.

> ## Example Use Case

Reviewer: ```suggestion
## Example use case
```

Author: Done

              env:
              - name: MASTER_ADDR
                value: "pytorch-workers-0-0.pytorch"
              - name: MASTER_PORT
                value: "3389"
              - name: RANK
                valueFrom:
                  fieldRef:
                    fieldPath: metadata.annotations['batch.kubernetes.io/job-completion-index']
              # Force python to not buffer output and write directly to stdout, so we can view training logs via `kubectl logs`.
              - name: PYTHONUNBUFFERED
                value: "0"
              command:
              - bash
              - -xc
              - |
                torchrun --rdzv_id=123 --nnodes=4 --nproc_per_node=1 --master_addr=$MASTER_ADDR --master_port=$MASTER_PORT --node_rank=$RANK mnist.py --epochs=1 --log-interval=1 
```

Leader-Worker

A popular paradigm for distributed workloads is the leader-worker design. Leader-worker entails having a leader pod, with a different template than workers, and worker pods. The leader must be started before the workers are started and if all the workers finish we can consider the workload complete and the overallworkload can be considered successful.

> a) Different pod templates for workers and leaders
> b) Success policies on the workers
> c) Startup policies

Reviewer: ```suggestion
1. Different pod templates for workers and leaders
1. Success policies on the workers
1. Startup policies
```

Author: Changed to numbered list

Example workload

```yaml
apiVersion: jobset.x-k8s.io/v1alpha2
kind: JobSet
metadata:
  name: success-policy
spec:
# We want to start our JobSet in order with leader starting first and then workers starting
  startupPolicy:
    startupPolicyOrder: InOrder
# We want to declare our JobSet successful if workers finish.
# If workers finish we should clean up the remaining replicatedJobs.
  successPolicy:
    operator: All
    targetReplicatedJobs:
    - workers
  replicatedJobs:
  - name: leader
    replicas: 1
    template:
      spec:
        # Set backoff limit to 0 so job will immediately fail if any pod fails.
        backoffLimit: 0 
        completions: 1
        parallelism: 1
        template:
          spec:
            containers:
            - name: leader
              image: bash:latest
              command:
              - bash
              - -xc
              - |
                sleep 10000

>   - name: workers

Reviewer 4: are there any tangible examples of this? I assume not if you're using bash here, but I find it odd this is abstract

Author: I agree, we should make the example more concrete.

Author: I think we can probably come up with a GPU based example and simply use fake node labels like "cloud.provider.com/gpu-label" instead of a vendor specific one

Reviewer 3: Maybe you can link to GKE related use cases as concrete examples? 

If @haircommander point is taking literal, we should show concrete usecases also.

Reviewer 3: What do you think about that @sftim? Should we try to keep the post self contained?

Author: Updated to use TPU multislice example

    replicas: 1
    template:
      spec:
        backoffLimit: 0 
        completions: 2
        parallelism: 2
        template:
          spec:
            containers:
            - name: worker
              image: bash:latest
              command:
              - bash
              - -xc
              - |
                sleep 10
```

> 

Reviewer 3: Let's end this with a getting involved and mention the batch working group..

Reviewer 3: Maybe use this as an example: https://kubernetes.io/blog/2022/10/04/introducing-kueue/#future-work-and-getting-involved

Author: Added

---
layout: blog
title: "Introducing JobSet"

> date: 2024-04-02

Reviewer 2: What publication date should we target?

Reviewer: Let's go for 2015-01-10 - does that work?

Reviewer 2: yeah, that is works great.

Author: Set publication date for 2025-01-13 since it was a few days til I had time to address these last comments, and want to give sufficient time for a final review.

Startup sequencing
: Some jobs require a specific start sequence of pods; sometimes the driver is expected to start first (like Ray or Spark), in other cases the workers are expected to be ready before starting the driver (like MPI).

JobSet aims to address those gaps using the Job API as a building block to build a richer API
for large-scale distributed HPC and ML use cases.

## How JobSet Works
JobSet models a distributed batch workload as a group of Kubernetes Jobs. This allows a user
to easily specify different pod templates for different distinct groups of pods (e.g. a leader,
workers, parameter servers, etc.). 

It uses the abstraction of a ReplicatedJob to manage child Jobs, where a ReplicatedJob is
essentially a Job Template with some desired number of Job replicas specified. This provides
a declarative way to easily create identical child-jobs to run on different islands of
accelerators, without resorting to scripting or Helm charts to generate many versions of the
same job but with different names.

> ![JobSet Architecture Diagram](jobset_diagram.svg "JobSet Architecture")

Reviewer: nit: using a [figure](https://gohugo.io/content-management/shortcodes/#figure) shortcode, with a caption, is even better. You can also set the `diagram-medium` class; I recommend doing that.

Author: Done, used figure with a caption

Some other key JobSet features which address the problems described above include:

Replicated Jobs

> : In modern data centers, hardware accelerators like GPUs and TPUs allocated

Reviewer 2: The headers look strange, should we move this line up and use a bold font for the header?

Reviewer: IMO this is the right markdown

If the style doesn't look right: we'll take a PR (or you can file an issue).

: Users can submit JobSets via [Kueue](https://kueue.sigs.k8s.io/) to oversubscribe their clusters, queue workloads to run as capacity becomes available, prevent
partial scheduling and deadlocks, enable multi-tenancy, and more.

## Example use case

### Distributed ML training on multiple TPU slices with Jax

The following example is a JobSet spec for running a TPU Multislice workload on
4 TPU v5e [slices](https://cloud.google.com/tpu/docs/system-architecture-tpu-vm#slices). To learn 
more about TPU concepts and terminology, please refer to these [docs](https://cloud.google.com/tpu/docs/system-architecture-tpu-vm).

This example uses [Jax](https://jax.readthedocs.io/en/latest/quickstart.html), an ML framework with native support for Just-In-Time (JIT)
compilation targeting TPU chips via [OpenXLA](https://github.com/openxla). However, you can also use 
[PyTorch/XLA](https://pytorch.org/xla/release/2.3/index.html) to do ML training on TPUs.

This example makes use of several JobSet features (both explicitly and implicitly) to support the unique scheduling requirements of
TPU multislice training out-of-the-box with very little configuration required by the user.

```yaml

> # Run a simple Jax workload on 

Reviewer: Did you mean
```suggestion
# Run a simple Jax workload on Google Cloud
```
?

Reviewer 2: @danielvegamyhre do you mind address this

Author: Sure I'll do a pass through the comments tonight

Author: Didn't get time today, will try to do it tomorrow

Author: Fixed

apiVersion: jobset.x-k8s.io/v1alpha2
kind: JobSet
metadata:
  name: multislice
  annotations:
    # Give each child Job exclusive usage of a TPU slice 

>     alpha.jobset.sigs.k8s.io/exclusive-topology: cloud.google.com/gke-nodepool

Reviewer: We should register (and, ideally, deprecate) this annotation before we publicize it.
The page to register it on is https://kubernetes.io/docs/reference/labels-annotations-taints/

---

It needs registering because although it uses a deprecated pattern (“alpha” in the key), it's inside an official Kubernetes namespace for annotations.

Reviewer: If WG Batch do deprecate this annotation and switch to something different, that's a nice but separate improvement.

We still need the original one registered if any released code inside https://github.com/kubernetes/ uses it, or previously used it.

Reviewer 3: I don't know if we should register alpha annotations/labels. Sounds like drift waiting to happen.

I don't know if most sig projects register the labels/annotations on that website.

Reviewer: Marking annotations as alpha or beta is deprecated; we should just use the plain form (eg `jobset.kubernetes.io/exclusive-topology`). AIUI if a SIG wants to mark something as experimental, and for that not to be registered, it needs to be outside `kubernetes.io` or `k8s.io`.

We should register this one, _or_ pick something different and blog about the updated annotation.

Reviewer 3: @danielvegamyhre WDYT? Could we have a post without this annotation for now? I know that changing these annotations could break users.

Reviewer: Separate from this article: if / when we switch to an annotation following the recommended patterns, we can still support the old annotation as a legacy thing

[and in that case, it's still polite to register it; registering an annotation is a short 8 to 12 line PR against k/website]

Reviewer: Here's an example of an annotation that's registered but not used any more
https://kubernetes.io/docs/reference/labels-annotations-taints/#service-alpha-kubernetes-io-tolerate-unready-endpoints-deprecated

Author: This annotation is required for the example to work. Could we just update the blog post if/when we migrate the annotation from alpha to the plain form?

Reviewer 3: I think that in order to publish the post with those annotations, we should register them in the k8s website. 

I created https://github.com/kubernetes-sigs/jobset/issues/639 for this reason. I don't see very many sig related projects with annotations. In fact, the only project I see where annotations are registered is node feature discovery (NFD).

Reviewer: That's a good way forward. It should not be much work.

Reviewer: > I don't know if most sig projects register the labels/annotations on that website.

SIG Architecture don't have the capacity to follow up every place where people haven't followed the project's policy.

Reviewer 2: Does registering include projects outside https://github.com/kubernetes/? this is a subproject not an upstream k8s api.

Reviewer: Sure, but as soon as we blog about it, that article makes the annotations look official.

I do think the SIG should fix it (ie, register the annotations). _Much_ simpler than requesting a policy change.

Reviewer: Imagine if a vendor, ACMECorp, had asked to publish a blog article, and they were using an unregistered annotation from the protected namespace. eg `kubernetes.io/use-acmecorp-extensions: auto`.

I'd be asking them to either change the article or change the product, for much the same reason. To me, being in-project doesn't count as an excuse for skipping the policy either.

Reviewer: @ahg-g if you're wondering how much work is involved in registering the annotation, it's a 10 line PR to update https://kubernetes.io/docs/reference/labels-annotations-taints/

Please consider that if writing >= 10 lines of objection.

Reviewer: Got it! The way I've done it, anything with an in-project origin can be registered (even if it shouldn't have been used; as soon as code ships that might use the annotation, we have to document it).

Reviewer: Registered means “make a record that it's already in use“; it's not like an API review - deliberately!

Author: > https://kubernetes.io/docs/reference/labels-annotations-taints/

I can do this

Reviewer 2: This is now addressed in https://github.com/kubernetes/website/pull/49217

custom distributed training orchestration solutions to fit their needs.

However, as distributed ML training techniques continue to evolve, existing Kubernetes primitives
do not adequately model them alone anymore. 

Furthermore, the landscape of Kubernetes distributed
training orchestration APIs has become fragmented, and each of the existing solutions in this
fragmented landscape has certain limitations that make it non-optimal for distributed ML training. 

For example, the KubeFlow training operator defines custom APIs for different frameworks
(e.g. PyTorchJob, TFJob, MPIJob, etc.); however, each of these job types are in fact a solution
fit specifically to the target framework, each with different semantics and behavior. 

On the other hand, the Job API fixed many gaps for running batch workloads, including Indexed
completion mode, higher scalability, Pod failure policies and Pod backoff policy to mention
a few of the most recent enhancements. However, running ML training and HPC workloads using
the upstream Job API requires extra orchestration to fill the following gaps:

Multi-template Pods

> : Most HPC or ML training jobs include more than one type of Pods. The different Pods are part of the same workload, but they need to run a different container, request different resources or have different failure policies. A common example is the driver-worker pattern.

Reviewer: In https://github.com/kubernetes/website/pull/45759#discussion_r1548544715 I mentioned that not wrapping them is easiest.

Job groups

> : Large scale training workloads span multiple network topologies, running across multiple racks for example. Such workloads are network latency sensitive, and aim to localize communication and minimize traffic crossing the higher-latency network links. To facilitate this, the workload needs to be split into groups of Pods each assigned to a network topology.

Reviewer: In https://github.com/kubernetes/website/pull/45759#discussion_r1548544715 I mentioned that not wrapping them is easiest.

Inter-Pod communication
: Create and manage the resources (e.g. [headless Services](/docs/concepts/services-networking/service/#headless-services)) necessary to establish communication between the Pods of a job.

Startup sequencing

> : Some jobs require a specific start sequence of pods; sometimes the driver is expected to start first (like Ray or Spark), in other cases the workers are expected to be ready before starting the driver (like MPI).

Reviewer: In https://github.com/kubernetes/website/pull/45759#discussion_r1548544715 I mentioned that not wrapping them is easiest.

and lifecycle management of the headless service enabling this. 

Configurable success policies
: JobSet has configurable success policies which target specific ReplicatedJobs, with operators
to target “Any” or “All” of their child jobs. For example, you can configure the JobSet to be marked
complete if and only if all pods that are part of the “worker” ReplicatedJob are completed.

Configurable failure policies
: JobSet has configurable failure policies which allow the user to specify a maximum number of
times the JobSet should be restarted in the event of a failure. If any job is marked failed,
the entire JobSet will be recreated, allowing the workload to resume from the last checkpoint.
When no failure policy is specified, if any job fails, the JobSet simply fails. 

Exclusive placement per topology domain
: JobSet allows users to express that child jobs have 1:1 exclusive assignment to a topology
domain, typically an accelerator island like a rack. For example, if the JobSet creates
two child jobs, then this feature will enforce that the pods of each child job will be
co-located on the same island, and that only one child job is allowed to schedule per island.

> This useful for scenarios where, for example, we want to use a data parallel distributed

Reviewer 2: ```suggestion
This is useful for scenarios where we want to use a data parallel distributed
```

Author: Updated with your suggestion but also rephrased to "distributed data parallel" to align with the PyTorch naming (see [DDP docs](https://pytorch.org/tutorials/intermediate/ddp_tutorial.html)).
